File: examples/visualization/shared/projection.mjs
/**
 * Shared GraphForge projection for Node visualization examples.
 *
 * Loads Mark Newman's karate-club GML through GraphForge's public Node API and
 * returns the same deterministic projection consumed by Cytoscape.js / Sigma.js.
 */

import { createHash } from "node:crypto";
import {
  existsSync,
  mkdirSync,
  readFileSync,
  writeFileSync,
} from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { createWriteStream } from "node:fs";
import { pipeline } from "node:stream/promises";
import { Readable } from "node:stream";
import { execFileSync } from "node:child_process";
import { tableFromIPC } from "apache-arrow";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = join(HERE, "..");
const MANIFEST_PATH = join(ROOT, "dataset", "MANIFEST.json");
const CONTRACT_PATH = join(HERE, "contract.json");
const DEFAULT_CACHE = join(ROOT, ".cache");

function loadJson(path) {
  return JSON.parse(readFileSync(path, "utf8"));
}

function sha256File(path) {
  const hash = createHash("sha256");
  hash.update(readFileSync(path));
  return hash.digest("hex");
}

async function download(url, dest) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Download failed (${response.status}): ${url}`);
  }
  await pipeline(Readable.fromWeb(response.body), createWriteStream(dest));
}

export async function fetchDataset(cacheDir = DEFAULT_CACHE, { force = false } = {}) {
  const manifest = loadJson(MANIFEST_PATH);
  mkdirSync(cacheDir, { recursive: true });
  const archivePath = join(cacheDir, manifest.archive.filename);
  const extractDir = join(cacheDir, "karate");
  const expected = manifest.archive.sha256;

  if (force || !existsSync(archivePath) || sha256File(archivePath) !== expected) {
    await download(manifest.source_url, archivePath);
  }
  const actual = sha256File(archivePath);
  if (actual !== expected) {
    throw new Error(
      `Archive checksum mismatch: expected ${expected}, got ${actual}`,
    );
  }

  mkdirSync(extractDir, { recursive: true });
  execFileSync("unzip", ["-o", archivePath, "-d", extractDir], {
    stdio: "ignore",
  });

  for (const [relative, memberExpected] of Object.entries(manifest.archive.members)) {
    const memberPath = join(extractDir, relative);
    if (!existsSync(memberPath)) {
      throw new Error(`Missing archive member: ${relative}`);
    }
    const memberActual = sha256File(memberPath);
    if (memberActual !== memberExpected) {
      throw new Error(
        `Member checksum mismatch for ${relative}: expected ${memberExpected}, got ${memberActual}`,
      );
    }
  }
  return extractDir;
}

export function parseGmlUndirectedEdges(gmlText) {
  const edges = new Set();
  const re = /edge\s*\[\s*source\s+(\d+)\s+target\s+(\d+)/g;
  let match;
  while ((match = re.exec(gmlText)) !== null) {
    const left = Number(match[1]);
    const right = Number(match[2]);
    if (left === right) continue;
    const a = Math.min(left, right);
    const b = Math.max(left, right);
    edges.add(`${a},${b}`);
  }
  return [...edges]
    .map((key) => key.split(",").map(Number))
    .sort((x, y) => x[0] - y[0] || x[1] - y[1]);
}

function resolveGraphForge() {
  const envPath = process.env.GRAPHFORGE_NODE_PATH;
  if (envPath) {
    return import(pathToFileURL(envPath).href);
  }
  // Prefer the in-repo binding during development.
  const localBinding = join(
    ROOT,
    "..",
    "..",
    "crates",
    "graphforge-bindings-node",
    "index.js",
  );
  if (existsSync(localBinding)) {
    return import(pathToFileURL(localBinding).href);
  }
  return import("@curatelabs/graphforge");
}

export async function buildGraphForge(datasetDir) {
  const [{ GraphForge }, manifest] = await Promise.all([
    resolveGraphForge(),
    Promise.resolve(loadJson(MANIFEST_PATH)),
  ]);
  const extractDir = datasetDir || (await fetchDataset());
  const gml = readFileSync(join(extractDir, "karate.gml"), "utf8");
  const edges = parseGmlUndirectedEdges(gml);
  if (edges.length !== manifest.graph.edge_count) {
    throw new Error(
      `Expected ${manifest.graph.edge_count} edges, found ${edges.length}`,
    );
  }

  const forge = new GraphForge();
  const handles = new Map();
  const [low, high] = manifest.graph.node_id_range;
  for (let clubId = low; clubId <= high; clubId += 1) {
    handles.set(
      clubId,
      forge.addNode(manifest.graph.node_label, {
        club_id: clubId,
        label: `M${clubId}`,
      }),
    );
  }
  if (handles.size !== manifest.graph.node_count) {
    throw new Error(
      `Expected ${manifest.graph.node_count} nodes, built ${handles.size}`,
    );
  }

  const rel = manifest.graph.relationship_type;
  for (const [source, target] of edges) {
    forge.addEdge(handles.get(source), rel, handles.get(target));
  }
  return forge;
}

function columnValues(table, name) {
  return [...table.getChild(name).toArray()].map((value) =>
    typeof value === "bigint" ? Number(value) : value,
  );
}

export async function project(forge, datasetDir) {
  const contract = loadJson(CONTRACT_PATH);
  const engine = forge || (await buildGraphForge(datasetDir));
  const nodeTable = tableFromIPC(engine.execute(contract.query.nodes));
  const edgeTable = tableFromIPC(engine.execute(contract.query.edges));

  const clubIds = columnValues(nodeTable, "club_id");
  const labels = columnValues(nodeTable, "label");
  const sources = columnValues(edgeTable, "source");
  const targets = columnValues(edgeTable, "target");

  return {
    projection_id: contract.projection_id,
    directed: contract.edge.directed,
    layout_seed: contract.layout.seed,
    style: {
      node_color: contract.node.color,
      node_size: contract.node.size,
      edge_color: contract.edge.color,
      edge_width: contract.edge.width,
    },
    nodes: clubIds.map((clubId, index) => ({
      id: Number(clubId),
      label: String(labels[index]),
      club_id: Number(clubId),
    })),
    edges: sources.map((source, index) => ({
      source: Number(source),
      target: Number(targets[index]),
    })),
  };
}

export function writeProjectionJson(path, projection) {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, `${JSON.stringify(projection, null, 2)}\n`, "utf8");
  return path;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  const out = join(ROOT, "output", "projection.json");
  const payload = await project();
  writeProjectionJson(out, payload);
  console.log(out);
}

File: examples/visualization/shared/projection.py
"""Shared GraphForge projection for visualization examples.

Loads Mark Newman's karate-club GML through GraphForge's public Python API and
returns a deterministic node/edge projection consumed by each library adapter.
"""

from __future__ import annotations

import json
from pathlib import Path
import re
import sys
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from dataset.fetch import fetch_dataset, load_manifest  # noqa: E402

CONTRACT_PATH = Path(__file__).resolve().parent / "contract.json"


def load_contract() -> dict[str, Any]:
    return json.loads(CONTRACT_PATH.read_text(encoding="utf-8"))


def parse_gml_undirected_edges(gml_text: str) -> list[tuple[int, int]]:
    """Parse undirected edges from Newman's karate.gml (source < target)."""
    edges: set[tuple[int, int]] = set()
    for match in re.finditer(
        r"edge\s*\[\s*source\s+(\d+)\s+target\s+(\d+)",
        gml_text,
    ):
        left, right = int(match.group(1)), int(match.group(2))
        if left == right:
            continue
        edges.add((left, right) if left < right else (right, left))
    return sorted(edges)


def build_graphforge(dataset_dir: Path | None = None):
    """Construct an in-memory GraphForge graph from the verified dataset."""
    from graphforge import GraphForge

    manifest = load_manifest()
    extract_dir = dataset_dir or fetch_dataset()
    gml_path = extract_dir / "karate.gml"
    edges = parse_gml_undirected_edges(gml_path.read_text(encoding="utf-8"))

    expected_nodes = manifest["graph"]["node_count"]
    expected_edges = manifest["graph"]["edge_count"]
    if len(edges) != expected_edges:
        raise RuntimeError(f"Expected {expected_edges} undirected edges, found {len(edges)}")

    forge = GraphForge()
    handles: dict[int, Any] = {}
    low, high = manifest["graph"]["node_id_range"]
    for club_id in range(low, high + 1):
        handles[club_id] = forge.add_node(
            manifest["graph"]["node_label"],
            club_id=club_id,
            label=f"M{club_id}",
        )
    if len(handles) != expected_nodes:
        raise RuntimeError(f"Expected {expected_nodes} nodes, built {len(handles)}")

    rel = manifest["graph"]["relationship_type"]
    for source, target in edges:
        forge.add_edge(handles[source], rel, handles[target])

    return forge


def project(forge=None, dataset_dir: Path | None = None) -> dict[str, Any]:
    """Return the shared projection dict via GraphForge public execute()."""
    contract = load_contract()
    engine = forge or build_graphforge(dataset_dir)

    nodes = engine.execute(contract["query"]["nodes"]).to_pylist()
    edges = engine.execute(contract["query"]["edges"]).to_pylist()

    projection = {
        "projection_id": contract["projection_id"],
        "directed": contract["edge"]["directed"],
        "layout_seed": contract["layout"]["seed"],
        "style": {
            "node_color": contract["node"]["color"],
            "node_size": contract["node"]["size"],
            "edge_color": contract["edge"]["color"],
            "edge_width": contract["edge"]["width"],
        },
        "nodes": [
            {
                "id": int(row["club_id"]),
                "label": str(row["label"]),
                "club_id": int(row["club_id"]),
            }
            for row in nodes
        ],
        "edges": [
            {
                "source": int(row["source"]),
                "target": int(row["target"]),
            }
            for row in edges
        ],
    }
    return projection


def write_projection_json(path: Path, projection: dict[str, Any] | None = None) -> Path:
    payload = projection or project()
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    return path


if __name__ == "__main__":
    out = ROOT / "output" / "projection.json"
    write_projection_json(out)
    print(out)
